# Analysis of Brandeis classes in pandas
Here we show how to use pandas to analyze the Brandeis course data stored in data/courses

Some questions we can try to answer
* which faculty have taught the most students over the Fall2004-Summer2021 interval
* which departments offer the most courses
* which departments have had the most students
* how have the enrollents by departments changed
* etc, as in PA01

We'll show how to use pivot tables and aggregation to answer some of these.

This notebook shows how to use aggregation with groupby, pivot tables, and how to create new columns by applying some function to other columns.


In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

First we read in the data from Fall 2004 to Summer 2021 with minimal info about each course

In [2]:
df = pd.read_csv('data/classes/courses.csv')
df

subj  num suf format         term  code                       inst  \
0      NEJS  188   A    LEC    Fall 2004  1043              Levy, Avigdor   
1      CHEM  121   A    LEC    Fall 2004  1043              Foxman, Bruce   
2      CHEM  130   A    LEC    Fall 2004  1043                Yu, Jinquan   
3      COMP    1   A    LEC    Fall 2004  1043             Ruesch, Gordon   
4       MUS  118   A    STD    Fall 2004  1043                  Ray, Mary   
...     ...  ...  ..    ...          ...   ...                        ...   
35201  BIOL   26   A    LEC  Summer 2021  1212  Kosinski-Collins, Melissa   
35202  ECON   28   B    LEC  Summer 2021  1212         AlMehdar, Mohammed   
35203  PHIL    1   A    LEC  Summer 2021  1212                Lupu, Peter   
35204  AAAS  160   B    LEC  Summer 2021  1212         Plumb, Betsy Loren   
35205   IGS  130   A    LEC  Summer 2021  1212            Campion, Selene   

                                                   title sec  enr  
0                               RISE/DECLINE:OTTOMAN EMP   1   32  
1                                INORGANIC CHEM I, LECS.   1   18  
2                               ADV ORG. CHEM: STRUCTURE   1   14  
3                                            COMPOSITION   3   11  
4                                       SENIOR RECITAL I   1    3  
...                                                  ...  ..  ...  
35201                                      Plant Biology   1    9  
35202                                 The Global Economy   1    4  
35203                         Introduction to Philosophy   1    3  
35204  If We Must Die: War and Military Service in Af...   1    3  
35205                                   Global Migration   1    1  

[35206 rows x 10 columns]

# Finding teachers with the most students
The easiest way to do this is with a groupby it shows the three stages of groupby
1. grouping of rows together based on the value in a column ('inst')
1. projection of those rows into a set of values (what value are we interested in) (enr)
1. aggregation of those values into a single value with some operator from arrays to numbers (sum)

In [8]:
#fac = df.groupby(['inst','subject'])['enr'].agg(sum)
#fac
fac = df.groupby(['inst','subj'])
fac = fac['enr'].agg([sum, np.mean, np.max,len])
fac = fac.sort_values(ascending=False,by='sum')
fac.head(40)



sum        mean  amax  len
inst                      subj                             
Coiner, Harry             ECON  8646  118.438356   203   73
Dolnik, Milos             CHEM  8354   31.288390    73  267
Kosinski-Collins, Melissa BIOL  6654   47.191489    87  141
Wright, Ellen             PSYC  4970   71.000000   193   70
Novack, Claudia           CHEM  4598   35.643411    88  129
Mascall, Kristen          CHEM  4577   52.011364   226   88
Teuber, Andreas           PHIL  4362   82.301887   199   53
Di Lillo, Antonella       COSI  4357   65.029851   149   67
Van Der Meid, J.          OFFC  4099  132.225806   214   31
Redenius, Scott           ECON  3827   61.725806   104   62
Hickey, Timothy           COSI  3427   46.945205   169   73
Canella, Alfonso          FIN   3379   35.197917    66   96
Pontrello, Jason          CHEM  3319   40.975309   160   81
Morris, James             BIOL  3305   63.557692   252   52
Cohen, Jacob              AMST  3210   74.651163   149   43
Simister, Neil            BIOL  3030   42.676056   105   71
Altman, Stuart            HS    3025   84.027778   187   36
Chilingerian, Jon         HS    3016   23.936508    75  126
Woodruff, Rachel          BIOL  2950   34.302326   302   86
Breen, Daniel             LGLS  2912   76.631579   190   38
Miara, Maria              BIOL  2885   32.415730   188   89
Lamb, Sarah               ANTH  2607   63.585366   206   41
Zimmerman, Grace          BUS   2582   30.738095    56   84
Wellenstein, Hermann      PHYS  2484   63.692308   169   39
McIntosh, Janet           ANTH  2442   55.500000   160   44
Fell, Richard             PHYS  2429   21.121739   126  115
Olson, Eric               HS    2354   27.057471    78   87
Reed Jr., Charles         BUS   2347   33.056338    59   71
Fournier, Stephen         HS    2323   24.712766    51   94
Godoy, Ricardo            HS    2310   19.913793    94  116
Kapelle, William          HIST  2213   37.508475    77   59
Morrison, Paul            ENG   2157   46.891304   135   46
Woll, Peter               POL   2153   43.060000    92   50
Flesch, William           ENG   2138   21.168317   134  101
Art, Robert               POL   2103   40.442308   144   52
Bayone, Edward            BUS   2070   31.846154    65   65
James, Heyward            HIST  2035   72.678571   148   28
Anderson, Brenda          FIN   1959   39.979592    63   49
Fellman, Gordon           SOC   1949   33.603448    85   58
Katz, Donald              NPSY  1931   96.550000   153   20

In [21]:

df[ df['inst']=='Perlman, Dan' ].groupby('title')['enr'].agg([sum, len, np.mean])

sum  len        mean
title                                               
ANIMAL BEHAVIOR                  94    4   23.500000
BIOLOGY OF FISHES                27    4    6.750000
BIOLOGY OF WHALES                56    4   14.000000
CLIMATE CHANGE BIODIVERSITY      33    1   33.000000
COASTAL ZONE MANAGEMENT           6    1    6.000000
CONSERVATION BIOLOGY            187    7   26.714286
ECOLOGY                         200    3   66.666667
EVOLUTIONARY ECOLOGY            108    3   36.000000
EXPERIENTIAL LEARNING PRACTICM   22    2   11.000000
FIELD BIOLOGY                    72    6   12.000000
FUNDAMENTALS ENVIRON. CHALLENG  622    6  103.666667
GIS&FIELD METHOD:N.E.LANDSCAPE   13    1   13.000000
INTRO TO MARINE MAMMALS           4    1    4.000000
MARITIME HISTORY OF N.E.          5    1    5.000000
TOPICS IN  ECOLOGY               47    1   47.000000
TOPICS IN ECOLOGY                19    1   19.000000
TROPICAL ECOLOGY                 21    1   21.000000
WATER RESRCES/PLAN &MGMT         15    1   15.000000
WETLANDS                          3    1    3.000000

## Sorting the Series
We can sort the values if we want (and here I am then reversing them by iterating with a step of -1 through the array) and we can use .head(N) and .tail(N) to look at either end of the sorted series.

In [4]:
fac = fac.sort_values(ascending=False)
fac.head(20)


TypeError: sort_values() missing 1 required positional argument: 'by'

In [ ]:
fac.head(1010).tail(10)

In [ ]:
df

# Finding departments (subj) with the most students per year
Let's do another example of aggregation, to sum the enrollments for each department over all years

In [ ]:
depts = df.groupby('subj')['enr'].agg(sum)
depts

It is more interesting to look at enrollments per year

In [ ]:
def div17(x):
    return x/17
depts = depts.apply(div17)
depts

In [ ]:
depts = depts.sort_values()
depts = depts[::-1]
depts.head(20)

# Restricting to a time interval
Let's do the same but just look for the past 4 years, i.e. from Fall2017 to Summer 2021 which are the codes 1173 to 1212.


In [ ]:
df2 = df[df['code']>=1193]
df2
#islater = df['code']>= 1173
#df[islater]

## aggregate by summing the enrollments for each subject

In [ ]:
depts = df2.groupby(['subj'])['enr'].agg(sum)
depts = depts.sort_values(ascending=False)
depts.head(20)
                     


# Time series
Now lets try to get time series data and plot the enrollments in a department by year.
Let's first add a new column giving the academic year, so Fall2004 and Spring 2005 and Summer2005 would all be 2004


In [ ]:
def code2year(code):
    ''' returns the academic year corresponding to a code'''
    code=str(code)
    year = int("20"+code[1:-1])
    if code[-1]!='3':
        year = year -1
    return year
print(code2year('1051'))



### the original data frame has no 'year' column

In [ ]:
df

### create a new column for the year
We can apply the code2year function to each of the integers in the 'code' column to get a new series,
and then assign that series to be a new column 'year' of the df dataframe

In [ ]:
df['year'] = df['code'].apply(code2year)
df

# Group by the year and subject
This creates a new Series with a multi-index containing both year and subject

In [ ]:
depts_enr = df.groupby(['subj','year'])['enr'].agg(sum)
depts_enr['COSI'].plot()
plt.axis([2004,2021,0,2500])
plt.grid()


In [ ]:
depts_enr = df.groupby(['year','subj'])['enr'].agg(sum)
depts_enr

# Reset to a single index
We can use the reset_index method to make the subj part of the index into a column


In [ ]:
depts_enr = depts_enr.reset_index(level=['subj'])
depts_enr

# Create a Pivot table of enrollments by department and year
Now we do the usual construction of a pivot table

In [ ]:
depts_pt =  pd.pivot_table(depts_enr, values='enr', index=["year"], columns=["subj"])
depts_pt

# Plot selected columns of the pivot table
We can now select a few columns and plot them!

In [ ]:
depts_pt[['BIOL','CHEM','COSI','MATH','PHYS','BCHM']].plot(figsize=(15,10))
plt.grid()
plt.title("STEM Enrollments from 2004-2020")

In [ ]:
depts_pt.plot(figsize=(15,10))
plt.legend([])

## Most popular course at Brandeis

In [ ]:
df

In [ ]:
df2 = df[df['code']>1042]
courses = df2.groupby(['title'])['enr'].agg(sum)
courses.sort_values(inplace=True,ascending=False)
courses.head(10)

In [ ]:
fac = df.groupby('title')
fac = fac['enr'].agg([sum])
fac = fac.sort_values(ascending=False, by='sum').head(10)
fac

In [ ]:
popcourses = df.groupby('title')['enr'].agg([sum]).sort_values(ascending=False, by='sum').head(10)
popcourses
